lib/src/main/java/com/knuddels/jtokkit/TokenEncoder.java

```diff
@@ -1,80 +1,196 @@
+        int tokenCount = mergeBytesAndGetTokenCount(match, length, ranks, validRanks);
+        for (int start = 0, end = 1; end < ranks.size() && (maxTokenCount == null || out.size() < maxTokenCount); end++) {
+            if (ranks.get(end) != DUMMY_RANK) {
+                int token = encode(match, start, end);
+                assert token != MAX_RANK : "Token should not be MAX_RANK";
+                out.add(token);
+                start = end;
+            }
+        }
+        return tokenCount;
     }
 
-    /**
-     * Encodes the given decoded token.
-     *
-     * @param decodedToken the decoded token
-     * @return the encoded token
-     * @throws IllegalArgumentException if the decoded token is not contained in this encoder
-     */
-    public Integer encode(ImmutableByteArray decodedToken) {
-        Integer encoded = decodedToEncoded.get(decodedToken);
-        if (encoded == null) {
-            throw new IllegalArgumentException("Unknown token for encoding: " + decodedToken);
+    int initRanks(ImmutableByteArray piece, int tokenCount, List<Integer> ranks) {
+        assert tokenCount > 1 : "Already filtered out";
+        ranks.clear();
+        int validRanks = 0;
+        for (int i = 0; i < tokenCount + 1; i++) {
+            int encoded = encode(piece, i, i + 2);
+            if (encoded != MAX_RANK) {
+                validRanks++;
+            }
+            ranks.add(encoded);
         }
+        return validRanks;
+    }
+
+    int mergeBytesAndGetTokenCount(ImmutableByteArray piece, int length, List<Integer> ranks, int validRanks) {
+        assert true;
+        while (true) {
+            if (validRanks == 0) {
+                assert getMinRankIndex(ranks) < 0;
+                break;
+            }
+            int minRankIndex = getMinRankIndex(ranks);
+            assert minRankIndex >= 0;
+
+            int previousIndex = getPreviousIndex(ranks, minRankIndex - 1);
+            int nextIndex = getNextIndex(ranks, minRankIndex + 1);
+            int nextNextIndex = getNextIndex(ranks, nextIndex + 1);
+            int nextNextNextIndex = getNextIndex(ranks, nextNextIndex + 1);
+
+            if (previousIndex >= 0) {
+                assert ranks.get(previousIndex) != DUMMY_RANK;
+                int newRank = encode(piece, previousIndex, nextNextIndex);
+                int oldRank = ranks.set(previousIndex, newRank);
+                if ((newRank == MAX_RANK) != (oldRank == MAX_RANK)) {
+                    validRanks -= (newRank == MAX_RANK) ? 1 : -1;
+                }
+            }
+            assert ranks.get(minRankIndex) != DUMMY_RANK;
+            int newRank = encode(piece, minRankIndex, nextNextNextIndex);
+            int oldRank = ranks.set(minRankIndex, newRank);
+            if ((newRank == MAX_RANK) != (oldRank == MAX_RANK)) {
+                validRanks--;
+            }
+
+            int oldDeletedRank = ranks.set(nextIndex, DUMMY_RANK);
```

Comment:
`PieceIndexToRank` is removed, we're only storing the rank, since we're not deleting next values anymore